### PR TITLE
Add initial APOGEE DR17 support

### PR DESCRIPTION
**Pull request by henrysky**
This PR adds a download and reformat script for APOGEE DR17.

Compared to galaxy spectra, this dataset also added pseudo-continuum normalized spectra on top of flux spectra and spectra here are always in rest-frame.

Related to issue https://github.com/AstroPile/FlatironMeeting2024/issues/12 and https://github.com/AstroPile/FlatironMeeting2024/issues/19

### CODE REVIEW

File: scripts/apogee/apogee.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    @classmethod
+    def _info(self):
+        """Defines the features available in this dataset."""
+        # Starting with all features common to spectral dataset
+        features = {
+            "spectrum": Sequence(
+                {
+                    "flux": Value(dtype="float32"),
+                    "ivar": Value(dtype="float32"),
+                    # "lsf_sigma": Value(dtype="float32"),
```
Comment (EiffL) [inline scripts/apogee/apogee.py:89]: is it that the lsf is constant? and/or the lsf is not very important here?

For uniformity, we can still have an `lsf_sigma` entry, but set it to a constant value. The main reason for doing that is that this way, a neural network built to support lsf information will also understand this type of data.

File: scripts/apogee/build_parent_sample.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        return fullfilename
+    else:
+        try:
+            urllib.request.urlretrieve(urlstr, fullfilename)
+            return 0
+        except urllib.error.HTTPError as emsg:
+            return 1  # error code
+
+
+def selection_fn(base_path, catalog):
```
Comment (EiffL) [inline scripts/apogee/build_parent_sample.py:100]: can you move this selection function to the top of the file, to make it really easy for people to find?

File: scripts/apogee/build_parent_sample.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        raw_ivar = np.zeros(8575)
+        mask_spec = np.zeros(8575)
+
+    # Load the combined spectra file
+    if os.path.exists(continuum_filename) and continuum_filename != 1:
+        hdus = fits.open(continuum_filename)
+        continuum_flux = hdus[1].data
+        continuum_ivar = 1 / hdus[2].data ** 2
+    else:
+        continuum_flux = np.zeros(8575)
```
Comment (EiffL) [inline scripts/apogee/build_parent_sample.py:141]: does this mean that we can have empty entries? Can we find those at the catalog level and remove them with some sort of selection?

**Review by EiffL — COMMENTED** [review]
Looking really good already! I have not yet tried to run it on my machine to see if it worked, just had a few questions first

File: scripts/apogee/build_parent_sample.py (hunk 3 above)
Comment (henrysky) [reply scripts/apogee/build_parent_sample.py]: Actually the selection function have filtered out those empty entries so I have removed these codes now

File: scripts/apogee/build_parent_sample.py (hunk 2 above)
Comment (henrysky) [reply scripts/apogee/build_parent_sample.py]: done

File: scripts/apogee/apogee.py (hunk 1 above)
Comment (henrysky) [reply scripts/apogee/apogee.py]: I have added a rough estimate on the lsf. lsf is not important here.

File: scripts/apogee/apogee.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    "radial_velocity",
+    "restframe",
+]
+
+# Features that correspond to ugriz fluxes
+_FLUX_FEATURES = []
+_BOOL_FEATURES = []
+
+
+class SDSS(datasets.GeneratorBasedBuilder):
```
Comment (EiffL) [inline scripts/apogee/apogee.py:66]:
```suggestion
class APOGEE(datasets.GeneratorBasedBuilder):
```

Comment (henrysky) [reply scripts/apogee/apogee.py]: I have changed the class name

**Comment by henrysky** [comment]
Thanks for the comments. I have pushed changes and seems to have fixed the issue. Looking at the failing CI test, it has failed at hsc dataset not APOGEE.

**Review by EiffL — APPROVED** [review]
This is great! Everything seems to work! Thanks @henrysky !

**Pull request merged**